### ISSUE TITLE
Browse file on remote

### ISSUE THREAD

**Issue by vermiculus** [magit/forge#91]
Optionally with selected region highlighted.

vermiculus/magithub#400

**Comment by tarsius** [comment]
> Browse file on remote

Yes, though I don't think that's urgent.  I prefer viewing files in
Emacs :-)

While I would probably not "browse" files much, I would still like
to be able to quickly get the corresponding url so that I can paste
them somewhere.  Luckily you have added support for that recently ;-)

> This will probably involve a new slot (or three) in `forge-repository`

But those will just be class-allocated `*-format` slots right?  So not
a big deal.

**Comment by tarsius** [comment]
Oh, and feel free to add issues such as this one, but please add the `enhancement` label right away.

Also if you do intend on something yourself, then assign it to yourself.  Hm, you probably are not allowed to do that yet.  Time to give you some permissions.

**Comment by Linuus** [comment]
Any status on this one? Would be a really great feature!

**Comment by achempion** [comment]
There is a package with the same name which does the same thing https://github.com/rmuslimov/browse-at-remote.

**Comment by Linuus** [comment]
> There is a package with the same name which does the same thing https://github.com/rmuslimov/browse-at-remote.

Yeah I know. There are multiple, I am using git-link myself: https://github.com/sshaw/git-link

It would just be nice to have it in forge/magit instead :)

**Comment by fredefox** [comment]
I created a snippet for this. I took inspiration for how `forge-browse-commit` was implemented. Here's what I came up with:
```
(defun forge-browse-buffer-file ()
  (interactive
   (browse-url
    (let
        ((rev (magit-get-current-branch))
         (repo (forge-get-repository 'stub))
         (file (file-relative-name buffer-file-name (projectile-project-root))))
      (forge--format repo "https://%h/%o/%n/blob/%r/%f"
                     `((?r . ,rev) (?f . ,file)))))))
```
It requires `projectile` for figuring out the project-relative path of currently visited file.

**Comment by Sevensidedmarble** [comment]
@fredefox's implementation is cool but `projectile-project-root` is a little janky for me. It would be nice if magit just kept a variable referencing the relative path (from the base git directory) to the current buffer. I would think it wouldn't be hard, but my elisp-fu kinda sucks.

**Comment by tarsius** [comment]
> It would be nice if magit just kept a variable referencing the relative path (from the base git directory) to the current buffer.

```elisp
(magit-file-relative-name buffer-file-name)
```

**Comment by raszi** [comment]
This feature would be really useful if you would like to send over a reference to a file or a highlighted couple of lines to somebody over chat or email.

**Comment by fredefox** [comment]
> This feature would be really useful if you would like to send over a reference to a file or a highlighted couple of lines to somebody over chat or email.
I suppose you could combine the snippet @tarsius shared with the snippet I wrote replacing `browse-url` with `kill-new`.

**Comment by bezirg** [comment]
> > This feature would be really useful if you would like to send over a reference to a file or a highlighted couple of lines to somebody over chat or email.
> 
> I suppose you could combine the snippet @tarsius shared with the snippet I wrote replacing `browse-url` with `kill-new`.

```lisp
(defun forge-browse-buffer-file ()
  (interactive)
  (kill-new
    (let
        ((rev (magit-get-current-branch))
         (repo (forge-get-repository 'stub))
         (file (magit-file-relative-name buffer-file-name)))
      (forge--format repo "https://%h/%o/%n/blob/%r/%f"
                     `((?r . ,rev) (?f . ,file))))))
```

**Comment by raszi** [comment]
Thank @bezirg I could understand and combine the two solution. Actually I am working on a PR to add this functionality with line handling as well.

The solution is working so far but I am struggling with [this issue](https://emacs.stackexchange.com/q/62023/16631).

**Comment by raszi** [comment]
I've created at PR #317

**Comment by umanwizard** [comment]
This version extends @fredefox 's to use the revision hash, rather than the branch name (so that the link is permanent), and adds highlighting from the region if applicable. (Only tested on GitHub so far)

```elisp
(defun btv-forge-browse-buffer-file ()
  (interactive)
  (browse-url
   (let
       ((rev (magit-rev-abbrev "HEAD"))
        (repo (forge-get-repository 'stub))
        (file (magit-file-relative-name buffer-file-name))
        (highlight
         (if
             (use-region-p)
             (let ((l1 (line-number-at-pos (region-beginning)))
                   (l2 (line-number-at-pos (- (region-end) 1))))
               (format "#L%d-L%d" l1 l2))
           ""
           )))
     (forge--format repo "https://%h/%o/%n/blob/%r/%f%L"
                    `((?r . ,rev) (?f . ,file) (?L . ,highlight))))))
```

**Comment by fredefox** [comment]
That's a very nice addition. I would probably replace `magit-rev-abbrev` with `magit-rev-parse`. Also I would suggest defining it as a helper `forge--buffer-file` so you can do e.g.:
```
(defun forge-browse-buffer-file ()
  (interactive)
  (browse-url (forge--buffer-file)))

(defun forge-buffer-file-as-kill ()
  (interactive)
  (kill-new (forge--buffer-file)))
```

**Comment by tarsius** [comment]
I've added `forge-browse-blob`, see it's docstring for more information. There is no default key binding. In file/blob visiting buffers `forge-browse` should behave the same and `forge-copy-url-at-point-as-kill` kills the blob url when invoked there.
